ocaml/xapi/message_forwarding.ml

```diff
@@ -6432,4 +6440,139 @@ functor
+        let client_fn session_id rpc =
+          Client.Observer.register ~rpc ~session_id ~self ~host
+        in
+        do_op_on ~__context ~host ~local_fn client_fn
+
+      let unregister ~__context ~self ~host =
+        info "Observer.unregister: self=%s" (observer_uuid ~__context self) ;
+        let local_fn = Local.Observer.unregister ~self ~host in
+        let client_fn session_id rpc =
+          Client.Observer.unregister ~rpc ~session_id ~self ~host
+        in
+        do_op_on ~__context ~host ~local_fn client_fn
+
+      let destroy ~__context ~self =
+        info "Observer.destroy: self=%s" (observer_uuid ~__context self) ;
+        let hosts = Db.Observer.get_hosts ~__context ~self in
+        let local_fn =
+          Local.Observer.register ~self ~host:(Helpers.get_localhost ~__context)
+        in
+        Xapi_observer.observed_hosts_of ~__context hosts
+        |> List.iter (fun host ->
+               do_op_on ~__context ~host ~local_fn (fun session_id rpc ->
+                   Client.Observer.unregister ~rpc ~session_id ~self ~host
+               )
+           ) ;
+        Local.Observer.destroy ~__context ~self
+
+      let set_hosts ~__context ~self ~value =
+        let uuid = observer_uuid ~__context self in
+        info "Observer.set_hosts: self=%s value=%s" uuid
+          (String.concat ";" (List.map (host_uuid ~__context) value)) ;
+        Local.Observer.set_hosts ~__context ~self ~value ;
+        let new_hosts =
+          RefSet.of_list (Xapi_observer.observed_hosts_of ~__context value)
+        in
+        let old_hosts =
+          RefSet.of_list
+            (Xapi_observer.observed_hosts_of ~__context
+               (Db.Observer.get_hosts ~__context ~self)
+            )
+        in
+        let to_add = RefSet.diff new_hosts old_hosts in
+        let to_remove = RefSet.diff old_hosts new_hosts in
+        RefSet.iter
+          (fun host ->
+            let local_fn = Local.Observer.unregister ~self ~host in
+            do_op_on ~__context ~host ~local_fn (fun session_id rpc ->
+                Client.Observer.unregister ~rpc ~session_id ~self ~host
+            )
+          )
+          to_remove ;
+        RefSet.iter
+          (fun host ->
+            let local_fn = Local.Observer.register ~self ~host in
+            do_op_on ~__context ~host ~local_fn (fun session_id rpc ->
+                Client.Observer.register ~rpc ~session_id ~self ~host
+            )
+          )
+          to_add
+
+      let set_enabled ~__context ~self ~value =
+        info "Observer.set_enabled: self=%s value=%B"
+          (observer_uuid ~__context self)
+          value ;
+        let local_fn = Local.Observer.set_enabled ~self ~value in
+        let client_fn session_id rpc =
+          Client.Observer.set_enabled ~rpc ~session_id ~self ~value
+        in
+        List.iter
+          (fun host -> do_op_on ~__context ~host ~local_fn client_fn)
+          (Db.Host.get_all ~__context)
```

Comment:
I suspect this call (an other below as well) need to be called on all members and lastly on the coordinator so the database is only set when there have been no errors, while rolling back changes to end up always in a consistent state. I doubt this call can easily end up in a partial state, so I don't think it's pressing for now

There's the function [`Xapi_pool_helpers.call_fn_on_slaves_then_master`](https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_pool_helpers.ml#L258) that doesn't do the rollback on error correctly and does not locally call the coordinator function. I think that having a function that does all this is something that would benefits other calls in xapi. Maybe for another PR that I can make

Reply:
That is a good idea and even reduced the code needed

Reply:
Resolved by [600baec](https://github.com/xapi-project/xen-api/pull/4966/commits/600baec4d018797587fb3a25856f7e49429a1f81)